tests/perf-system/generator/generator.py

```diff
@@ -12,59 +12,46 @@
 
 class Messages:
     def __init__(self):
-        self.df = pd.DataFrame(columns=["messageID", "request"])
+        self.requests = []
 
     def append(
         self,
-        host,
         path,
         verb,
-        request_type="HTTP/1.1",
+        http_version="HTTP/1.1",
         content_type="application/json",
-        additional_headers="",
-        data="",
-        iterations=1,
+        additional_headers=None,
+        body=bytes(),
     ):
         """
-        Create a new df with the contents specified by the arguments,
-        append it to self.df and return the new df
+        Serialise HTTP request specified by the arguments, and
+        append it to self.requests
         """
-        batch_df = pd.DataFrame(columns=["messageID", "request"])
-        data_headers = b"\r\n"
-        if len(additional_headers) > 0:
-            additional_headers += "\r\n"
-        if len(data) > 0:
-            if isinstance(data, str):
-                data = data.encode("ascii")
-            data_headers = (f"content-length: {len(data)}\r\n\r\n").encode(
-                "ascii"
-            ) + data
 
-        df_size = len(self.df.index)
+        headers = {}
+        if additional_headers is not None:
+            headers.update({k.lower(): v for k, v in additional_headers.items()})
 
-        for ind in range(iterations):
-            batch_df.loc[ind] = [
-                str(ind + df_size),
-                (
-                    verb.upper()
-                    + " "
-                    + path
-                    + " "
-                    + request_type
-                    + "\r\n"
-                    + "host: "
-                    + host
-                    + "\r\n"
-                    + additional_headers
-                    + "content-type: "
-                    + content_type.lower()
-                    + "\r\n"
-                ).encode("ascii")
-                + data_headers,
-            ]
+        # Insert content-length, and content-type headers, if they're not already present
+        if "content-length" not in headers:
+            headers["content-length"] = str(len(body))
+        if "content-type" not in headers and content_type is not None:
+            headers["content-type"] = content_type
 
-        self.df = pd.concat([self.df, batch_df])
-        return batch_df
+        # Convert body to bytes if we were given a string
+        if type(body) == str:
+            body = body.encode("utf-8")
+
+        request_line = f"{verb.upper()} {path} {http_version}"
+        headers_string = "\r\n".join(f"{k}: {v}" for k, v in headers.items())
+        serialised_request = (
+            f"{request_line}\r\n{headers_string}\r\n\r\n".encode("ascii") + body
+        )
+
+        self.requests.append(
+            {"messageID": str(len(self.requests)), "request": serialised_request}
+        )

```

Comment:
Appending single rows to a pandas DataFrame is slow, and afaict there isn't a fast way to do it. We don't actually need a DataFrame here (its just what `fastparquet` wants to write the file from), so we instead use a Python list and only convert to DataFrame when writing the file.

I've also removed the iterations option and the `batch_df` intermediate, because they were unused. We could also refactor the header construction here - we're doing a lot of work to format an identical prefix to every request (request line and common headers), which could be done once. But this already takes the generation time from several minutes to under a second, so I think that additional refactor is unnecessary.